Source/WebCore/dom/Element.cpp

```diff
@@ -2352,26 +2352,48 @@ static RefPtr<Element> getElementByIdIncludingDisconnected(const Element& startE
     return nullptr;
 }
 
-RefPtr<Element> Element::getElementAttribute(const QualifiedName& attributeName) const
+RefPtr<Element> Element::getElementForAttributeInternal(const QualifiedName& attributeName) const
 {
-    ASSERT(isElementReflectionAttribute(document().settings(), attributeName));
+    RefPtr<Element> element = nullptr;
+    bool hasExplicitlySetElement = false;
 
     if (auto* map = explicitlySetAttrElementsMapIfExists()) {
         auto it = map->find(attributeName);
         if (it != map->end()) {
             ASSERT(it->value.size() == 1);
-            RefPtr element = it->value[0].get();
-            if (element && isDescendantOrShadowDescendantOf(element->rootNode()))
-                return element;
-            return nullptr;
+            hasExplicitlySetElement = true;
+            RefPtr explicitlySetElement = it->value[0].get();
+            if (explicitlySetElement && isDescendantOrShadowDescendantOf(explicitlySetElement->rootNode()))
+                element = explicitlySetElement;
         }
     }
+    
+    if (!hasExplicitlySetElement) {
+    const AtomString& id = getAttribute(attributeName);
+        element = getElementByIdInternalIncludingDisconnected(*this, id);
+    }
+
+    if (element == nullptr)
+        return nullptr;
+
+    return element->deepShadowRootReferenceTargetOrSelf();
+}
+
+RefPtr<Element> Element::getElementAttributeForBindings(const QualifiedName& attributeName) const
+{
+    ASSERT(isElementReflectionAttribute(document().settings(), attributeName));
+    RefPtr<Element> element = getElementForAttributeInternal(attributeName);
 
-    auto id = getAttribute(attributeName);
-    if (id.isNull())
+    if (!element)
         return nullptr;
 
-    return getElementByIdIncludingDisconnected(*this, id);
+    if (document().settings().shadowRootReferenceTargetEnabled()) {
+        Ref<Node> retargeted = treeScope().retargetToScope(*element);
+        ASSERT(retargeted->isElementNode());
+        return dynamicDowncast<Element>(retargeted);
+    }
```

Comment:
One option would be to use `downcast`, which inherently does a `RELEASE_ASSERT(is<T>(...))`. That would let us make this more terse overall, and remove the curly-braces:

```cpp
if (document().settings().shadowRootReferenceTargetEnabled())
    return downcast<Element>(treeScope().retargetToScope(*element));
```